pkg/aws/policies.go

```
@@ -0,0 +1,215 @@
+	// When you specify more than one principal in the element, you grant permissions to each principal.
+	AWS []string `json:"AWS"`
+}
+
+func (c *awsClient) EnsureRole(name string, policy string, tagList map[string]string) (string, error) {
+	output, err := c.iamClient.GetRole(&iam.GetRoleInput{
+		RoleName: aws.String(name),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeNoSuchEntityException:
+				return c.createRole(name, policy, tagList)
+			default:
+				return "", err
+			}
+		}
+	}
+	role := output.Role
+	return aws.StringValue(role.Arn), nil
+}
+
+func (c *awsClient) createRole(name string, policy string, tagList map[string]string) (string, error) {
+	output, err := c.iamClient.CreateRole(&iam.CreateRoleInput{
+		RoleName:                 aws.String(name),
+		AssumeRolePolicyDocument: aws.String(policy),
+		Tags:                     getTags(tagList),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeEntityAlreadyExistsException:
+				return "", nil
+			}
+		}
+		return "", err
+	}
+	role := output.Role
+	return aws.StringValue(role.Arn), nil
+}
+
+func (c *awsClient) PutRolePolicy(roleName string, policyName string, policy string) error {
+	_, err := c.iamClient.PutRolePolicy(&iam.PutRolePolicyInput{
+		RoleName:       aws.String(roleName),
+		PolicyName:     aws.String(policyName),
+		PolicyDocument: aws.String(policy),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *awsClient) EnsurePolicy(name string, document string, tagList map[string]string) error {
+	_, err := c.iamClient.CreatePolicy(&iam.CreatePolicyInput{
+		PolicyName:     aws.String(name),
+		PolicyDocument: aws.String(document),
+		Tags:           getTags(tagList),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeEntityAlreadyExistsException:
+				return nil
+			}
+		}
+		return err
+	}
```

Is it a conscious decision to not return an error if the policy already exists? How would this work when a customer upgrades from `4.7` to `4.8` for instance? Applying the `4.8` policies over the `4.7` policies for instance results in no change even though the policy for `4.8` adds a permission `elasticloadbalancing:DeregisterTargets` which is silently dropped in this case.

---

Good question. I'll add a way to not just ensure the role exists, but make sure it's the right version. Maybe I can add it in a follow-up PR?

---

My concern is we push this release and we're close to the 4.8 upgrade. It should be easy enough to validate the `rosa_openshift_version` tag and return an error if it doesn't match? If its a bigger change then ok let's do it in a separate PR.

---

@vkareh lets merge this one and call out the risk before we cut a release.